helpers/log_context.py

```diff
@@ -0,0 +1,83 @@
+import contextvars
+import logging
+from dataclasses import dataclass, replace
+
+from database.models.core import Commit, Repository
+
+log = logging.getLogger("log_context")
+
+
+@dataclass
+class LogContext:
+    """
+    Class containing all the information we may want to add in logs and metrics.
+    """
+
+    task_name: str = "???"
+    task_id: str = "???"
+
+    _populated_from_db = False
+    owner_id: int | None = None
+    repo_id: int | None = None
+    commit_sha: int | None = None
+    commit_id: int | None = None
+
+    def populate_from_sqlalchemy(self, dbsession):
+        """
+        Attempt to use the information we have to fill in other context fields. For
+        example, if we have `self.repo_id` but not `self.owner_id`, we can look up
+        the latter in the database.
+
+        Ignore exceptions; no need to fail a task for a missing context field.
+        """
+        if self._populated_from_db:
+            return
+
+        try:
+            if self.repo_id:
```

Comment:
So this approach relies on for sure having the repo_id? I suppose you couldn't really find the commit/repo from an owner id, although you could find a repo and owner from a commit info

Reply:
`LogContext` doesn't care if you don't even have the repo id. it just figures out whatever it can with whatever it has
- `owner_id` is not enough to figure out repo or commit
- `repo_id` is enough to find owner but not commit
- `repo_id` + `commit_sha` is enough to find owner + commit
- `commit_id` on its own is enough to find everything, but we never have it on its own so that case was never implemented
- `commit_sha` on its own is not enough to find anything because forks/mirrors will each have their own copy of the same SHAs